### PR TITLE
initial_commit

### PR DESCRIPTION
**Pull request by naili-m**
initial version for review

### CODE REVIEW

File: juniper-junos/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+validators >= 0.18.2
```
Comment (cs-prerna-joshi) [inline juniper-junos/requirements.txt:1]:
```suggestion
validators
```

Comment (cs-prerna-joshi) [reply juniper-junos/requirements.txt:1]: please do not freeze the versions of python lib.

File: juniper-junos/info.json [hunk 2]
```diff
@@ -0,0 +1,367 @@
+{
+    "name": "juniper-junos",
+    "version": "1.0.0",
+    "label": "Juniper JunOS",
+    "description": "Provides JunOS REST API Integration covering Juniper MX, PTX, QFX, T and SRX Series platforms",
+    "publisher": "",
```
Comment (cs-prerna-joshi) [inline juniper-junos/info.json]:
```suggestion
    "publisher": "Fortinet",
```

File: juniper-junos/info.json [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    "icon_small_name": "junos_small.png",
+    "icon_large_name": "junos_large.png",
+    "category": "Sandbox",
```
Comment (cs-prerna-joshi) [inline juniper-junos/info.json]:
```suggestion
    "category": "Firewall and Network Protection",
```

File: juniper-junos/info.json [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                    "placeholder": "get-interface-information",
+                    "tooltip": "if the command is not in the list above you can use a custom one here. To get the exact command syntax use this example on JunOS: show route|display xml rpc"
+                },
+                {
+                    "title": "Method Parameters",
+                    "name": "method_params",
+                    "required": false,
+                    "editable": true,
+                    "visible": true,
+                    "type": "text",
```
Comment (cs-prerna-joshi) [inline juniper-junos/info.json]:
```suggestion
                    "type": "json",
```

File: juniper-junos/info.json [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                    "visible": true,
+                    "editable": true,
+                    "required": true,
+                    "value": "",
+                    "placeholder": "10.1.3.1,2001:0db8:0000:0000:0000:8a2e:0370:7334",
+                    "tooltip": "IPv4 or IPv6 Address or Addresses (in CSV) to delete from the prefix list"
+                }
+            ],
+            "enabled": true,
+            "output_schema": ""
```
Comment (cs-prerna-joshi) [inline juniper-junos/info.json]: if output is available please add the output schema for actions.

File: juniper-junos/operations.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
+import requests, logging, json
+from requests.auth import HTTPBasicAuth
+from connectors.core.connector import get_logger, ConnectorError
+from .utils import *
+from .constants import *
+from requests_toolbelt.utils import dump
+
+logger = get_logger('juniper_junos')
+logger.setLevel(logging.DEBUG) #Comment this for prod
```
Comment (cs-prerna-joshi) [inline juniper-junos/operations.py]: can be removed.

File: juniper-junos/operations.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        if api_response.ok:
+            try:
+                return api_response.json()
+            except:
+                return api_response.content.decode('utf-8')
+        else:
+            logger.info('Fail To request API {0} response is : {1}'.format(str(url), str(api_response.content)))
+            raise ConnectorError('Fail To request API {0} response is : {1}'.format(str(url), str(api_response.content)))
+    except Exception as Err:
+        logger.info('API Request Error: {0}'.format(str(Err)))
```
Comment (cs-prerna-joshi) [inline juniper-junos/operations.py]:
```suggestion
        logger.error('API Request Error: {0}'.format(str(Err)))
```

File: juniper-junos/operations.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+def get_address_set(config,params):
+    '''
+    Returns address set set content or items count
+    :param config: Connector's config params
+    :param params.address_set: name of the address set to return
+    :param params.get_count: bool : is True returns nothing but items count
+    :return: address set data or count
+    '''
+    address_set = get_input(params,'address_set',str)
```
Comment (cs-prerna-joshi) [inline juniper-junos/operations.py]: you can skip input validations.

File: juniper-junos/operations.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    HTTP_HEADERS.update({'Accept':'application/xml','Content-Type': 'application/xml'})
+    payload = parse_address_set_params(params)
+    payload = CONFIG_COMMIT.format(payload)
+    logger.debug(payload)
+    try:
+        response = _api_request('post', '', config, payload=payload)
+        logger.debug(response)
+        return parse_config_xml_response(response)
+    except Exception as Err:
+        logger.info('_add_delete_address Error: {0}'.format(str(Err)))
```
Comment (cs-prerna-joshi) [inline juniper-junos/operations.py]:
```suggestion
        logger.error('_add_delete_address Error: {0}'.format(str(Err)))
```

File: juniper-junos/connector.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
+from connectors.core.connector import Connector, get_logger, ConnectorError
```
Comment (cs-sneha-bharote) [inline juniper-junos/connector.py:7]: Add below copyright code in python files.
""" Copyright start
  Copyright (C) 2008 - 2022 Fortinet Inc.
  All rights reserved.
  FORTINET CONFIDENTIAL & FORTINET PROPRIETARY SOURCE CODE
  Copyright end """

File: juniper-junos/info.json [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+            },
+            {
+                "title": "Username",
+                "required": true,
+                "editable": true,
+                "visible": true,
+                "type": "text",
+                "name": "username",
+                "placeholder": "root",
+                "description": "JunOS Username"
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: "description": "Username to access the Juniper JunOS server to which you will connect and perform the automated operations."

File: juniper-junos/info.json [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+            },
+            {
+                "title": "Password",
+                "required": true,
+                "editable": true,
+                "visible": true,
+                "type": "password",
+                "name": "password",
+                "description": "JunOS Password"
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Update config params descriptions.
"description": "Password to access the Juniper JunOS server to which you will connect and perform the automated operations."

File: juniper-junos/info.json [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+            }
+        ]
+    },
+    "operations": [
+        {
+            "operation": "op_action",
+            "title": "Run Operation Command",
+            "category": "information",
+            "annotation": "op_action",
+            "description": "Run JunOS CLI in operation mode to fetch data such as routing table, interfaces system info...etc",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Run a CLI command in operation mode on the JunOS console based on the commands and other input parameters you have specified to fetch data such as routing table, interfaces system info etc.

File: juniper-junos/info.json [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+                "type": "checkbox",
+                "value": true,
+                "description": "Specifies whether the SSL certificate for JunOS is to be verified or not.By default, this option is set as True."
+            }
+        ]
+    },
+    "operations": [
+        {
+            "operation": "op_action",
+            "title": "Run Operation Command",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Update the title. The title should be "Run Command" or "Run RPC Command"

File: juniper-junos/info.json [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+            "annotation": "op_action",
+            "description": "Run JunOS CLI in operation mode to fetch data such as routing table, interfaces system info...etc",
+            "parameters": [
+                {
+                    "title": "Method",
+                    "name": "method",
+                    "type": "select",
+                    "visible": true,
+                    "editable": true,
+                    "required": true,
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Method parameter should be optional if the user wants to run only Custom Method. You can give the option to the user to run predefined Method or Custom Method using onchange.

File: juniper-junos/info.json [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+                {
+                    "title": "Custom Method",
+                    "name": "custom_method",
+                    "required": false,
+                    "editable": true,
+                    "visible": true,
+                    "type": "text",
+                    "placeholder": "get-interface-information",
+                    "tooltip": "if the command is not in the list above you can use a custom one here. To get the exact command syntax use this example on JunOS: show route|display xml rpc"
+                },
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json:170]: In this action, you can add one more parameter "Result Format" which specifies the format for the output as XML, JSON, or plain text.

File: juniper-junos/info.json [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+            ],
+            "enabled": true,
+            "output_schema": ""
+        },
+        {
+            "operation": "config_action",
+            "title": "Run Configuration Command",
+            "category": "Configuration",
+            "annotation": "config_action",
+            "description": "Update JunOS Configuration",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Update the action description.

File: juniper-junos/info.json [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+                {
+                    "title": "Request Payload",
+                    "name": "request_payload",
+                    "type": "textarea",
+                    "visible": true,
+                    "editable": true,
+                    "required": true,
+                    "value": "",
+                    "placeholder": "<lock-configuration/><load-configuration><configuration><system><hostname>NewHostName</hostname></system></configuration></load-configuration><commit/>",
+                    "tooltip": "XML Payload. Check JunOS NETCONF for the exact syntax"
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Add a description of the Request Payload parameter. In that mention referral link of the payload format.

File: juniper-junos/info.json [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+            "parameters": [
+                {
+                    "title": "Address Set",
+                    "name": "address_set",
+                    "type": "text",
+                    "visible": true,
+                    "editable": true,
+                    "required": true,
+                    "value": "",
+                    "placeholder": "BlockedIPsAddressSet",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: is "BlockedIPsAddressSet" sample value of Address set?  Add description for all params
 "placeholder": "eg. BlockedIPsAddressSet",

File: juniper-junos/info.json [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+            ],
+            "enabled": true,
+            "output_schema": ""
+        },
+        {
+            "operation": "add_to_address_set",
+            "title": "Add an Object to Global Address Set",
+            "category": "Configuration",
+            "annotation": "add_to_address_book",
+            "description": "Add and IP address, an FQDN or a whildcard to an address set on the Global address book. 1024 entries Max",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Mention max limit in  "Object(s) To Add" param description.

File: juniper-junos/info.json [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+                    "tooltip": "Name of the address set"
+                },
+                {
+                    "title": "Object Types",
+                    "name": "object_type",
+                    "type": "select",
+                    "visible": true,
+                    "editable": true,
+                    "required": true,
+                    "options": [
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Specify user understandable on UI and map UI value with backend values.
Eg. "IP" mapped with "ip-prefix"

File: juniper-junos/info.json [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+                    "placeholder": "10.1.3.1",
+                    "tooltip": "IP address, an FQDN or a wildcard to add, for multiple entries use CSV format such as host1.domain.com,host2.domain.com if the type is dns-name"
+                }
+            ],
+            "enabled": true,
+            "output_schema": ""
+        },
+        {
+            "operation": "delete_from_address_set",
+            "title": "Delete Object from Global Address Set",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Other than the global address set present in JunOs? if not then we can make action names as:
Add Object, Delete Object.

File: juniper-junos/info.json [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+                    "required": true,
+                    "options": [
+                        "ip-prefix",
+                        "dns-name",
+                        "wildcard-address"
+                    ],
+                    "value": "ip-prefix",
+                    "tooltip": "Type of the object(s) to delete, only one type is supported at a time"
+                },
+                {
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Update as per above mentioned.

File: juniper-junos/info.json [hunk 24]
```diff
@@ -0,0 +1,367 @@
...
+            ],
+            "enabled": true,
+            "output_schema": ""
+        },
+        {
+            "operation": "get_prefix_list",
+            "title": "Get Prefix List",
+            "category": "Configuration",
+            "annotation": "get_prefix_list",
+            "description": "Get Prefix List entries",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: update description

File: juniper-junos/utils.py [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+    Parses JunOS configuration action response and extracts its XML, fixes it and converts it to JSON
+    :param response: Raw server response
+    :return: JSON formatted response
+    '''
+    xml_response = re.findall(r'\<.*\>', response)
+    if len(xml_response) > 0:
+        return_data = '<response>{0}</response>'.format(''.join(xml_response).replace('\\', ''))
+        return xmltodict.parse(return_data)
+
+def parse_op_xml_response(response_data):
```
Comment (cs-sneha-bharote) [inline juniper-junos/utils.py:139]: If user passes the output response format as JSON then no need to convert this XML to JSON. The response will come in json. please check this once.

File: juniper-junos/utils.py [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+    else:
+        if ret_val == {} or ret_val == [] or ret_val == 0:
+            return ret_val
+        return None
+
+def get_config(config):
+    '''
+    Returns JunOS config attributes
+    '''
+    device_url = config.get('device_url')[:-1] if config.get('device_url')[-1] == '/' else config.get('device_url')
```
Comment (cs-sneha-bharote) [inline juniper-junos/utils.py]: use strip

File: juniper-junos/utils.py [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+    if xml_response:
+        return xmltodict.parse(xml_response)
+
+def get_input(params, key, type):
+    '''
+    Validates user's input
+    '''
+    ret_val = params.get(key, None)
+    if ret_val:
+        if isinstance(ret_val, bytes):
```
Comment (cs-sneha-bharote) [inline juniper-junos/utils.py]: No need to do input value type validation.

File: juniper-junos/info.json [hunk 28]
```diff
@@ -0,0 +1,367 @@
...
+                "required": false,
+                "editable": true,
+                "visible": true,
+                "type": "checkbox",
+                "value": true,
+                "description": "Specifies whether the SSL certificate for JunOS is to be verified or not.By default, this option is set as True."
+            }
+        ]
+    },
+    "operations": [
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json]: Add playbooks for all actions.

File: juniper-junos/requirements.txt (hunk 1 above)
Comment (naili-m) [reply juniper-junos/requirements.txt:1]: This is mandatory for the default validators (0.11 shipped with 7.0.2) version doesn't include cidr functions. should not be a problem in the future for it is >= (superior or equal)

File: juniper-junos/utils.py (hunk 25 above)
Comment (naili-m) [reply juniper-junos/utils.py:139]: for all POST actions device response is always a malformed xml, no JSON despite the doc (their API bug)

File: juniper-junos/info.json [hunk 29]
```diff
@@ -0,0 +1,496 @@
+{
+    "name":"juniper-junos",
+    "version":"1.0.0",
+    "label":"Juniper JunOS",
+    "description":"Provides JunOS REST API Integration covering Juniper MX, PTX, QFX, T and SRX Series platforms",
+    "publisher":"Fortinet CSE",
```
Comment (cs-sneha-bharote) [inline juniper-junos/info.json:6]: "publisher":"Fortinet"
remove CSE.

File: juniper-junos/utils.py [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+    '''
+    xml_response = re.search(r'\<data.*data\>', response_data, re.DOTALL).group(0)
+    if xml_response:
+        return xmltodict.parse(xml_response)
+
+def get_config(config):
+    '''
+    Returns JunOS config attributes
+    '''
+    device_url = config.get('device_url')[:-1] if config.get('device_url')[-1] == '/' else config.get('device_url')
```
Comment (cs-sneha-bharote) [inline juniper-junos/utils.py]: remove this line.

File: juniper-junos/utils.py [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+    device_url = device_url.strip()
```
Comment (cs-sneha-bharote) [inline juniper-junos/utils.py]: device_url = device_url.strip('/')

File: juniper-junos/playbooks/playbooks.json [hunk 32]
```diff
@@ -0,0 +1,832 @@
+{
+  "type": "workflow_collections",
+  "data": [
+    {
+      "@type": "WorkflowCollection",
+      "name": "Sample - juniper-junos - 1.0.0",
+      "description": "",
```
Comment (cs-sneha-bharote) [inline juniper-junos/playbooks/playbooks.json]: add PB description as mentioned in info.json

File: juniper-junos/info.json (hunk 2 above)
Comment (naili-m) [reply juniper-junos/info.json]: Fortinet CSE is what has been agreed with Pooja, check please.

**Review by cs-sneha-bharote — APPROVED** [review]
LGTM.

**Pull request merged**